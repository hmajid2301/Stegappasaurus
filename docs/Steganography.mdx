---
name: Steganography
route: /steganography
---

# Steganography

The steganography code is the core logic of this app, as this whole app revolves around encoding and decoding images. So it will not shock you to know that the most complicated
part of the application is the code related to steganography.

## FAQ

### What is steganography?

Steganography is the art/science of hiding data in plain sight. In the context of this app, it
allows you to hide messages (text) within images. You do this by encoding the image with the
text data. You can then share the image with other people and they can decode the image
(retrieve the original message) using this app.

### Is it the same as encryption?

Not quite, encryption involves obscuring information whereas steganography is concerned with hiding
it. If you have an encrypted message you cannot retrieve the original message.
Whereas with steganography if you know where to look you can retrieve the original message, it is
not obscured in any way. However, with encryption, it is obvious you are trying to hide something
but with steganography you can hide your message inside of an "innocent" looking image, with a
third party without anyone noticing. You can combine the two "techniques", encrypt your data
before encoding it into an image, then simply decrypt the message after decoding it from the
image for added security. This involves a shared secret "password" between you and the person
you are sending the image to.

### How do images work?

Images are made of pixels, an image of say 1920 by 1080 pixels has `2073600` pixels.
Each pixel in the image has three components red, green and blue (RGB). Each component
takes up one byte of data i.e. it can range from `0 - 255` (decimal) or `10 - 11111111` (binary). When we get the image data we get each RGB component, so if an image was a white box
say the image data would look like.

_NOTE_: Images also have an alpha channel but we don't change this we leave it as `255`.

```
[255, 255, 255, 255, 255, 255]
  R    G    B    R    G    B
```

## Kotlin - Bitmap

The native Android code is written using Kotlin, it currently has two functions.
The Android code related to the steganography resides within

- `android/app/src/main/java/com/stegappasaurus/bitmap/`

### BitmapModule.kt

The module has two main functions it needs to do.

#### Get Pixel Data

The get pixels function opens an image and returns the pixels data of that image. For each pixel
we get a red byte, a green byte and a blue byte. So for each pixel, we can encode 3 bits into it,
and therefore also decode three bits from it. You don't get the all of the data from the image
you specify which pixel to start at and which pixel to end at.

#### Set Pixel Data

The set pixels function edits the pixel data of the image and then saves this image locally on the
device in a steganography folder. This function is only used when we are encoding an image.

## Typescript - Steganography

The Typescript (core) steganography code resides within

- `src/actions/Steganography`

### Steganography.ts

This is the class that the React components will call to encode/decode an image.

#### Encoding

All encoding algorithms work as follows

```typescript
* // Input: image_path: String, message: String, metadata: Object
* // Output: encoded_image_path: String

* const compressed_message: String = compress_message(message)
* const compressed_binary_message: String = convert_message_to_bits(compressed_message)
* const binary_metadata: String = convert_metadata_to_bits(metadata)
* const binary_message: String = binary_metadata + compressed_binary_message
* const image_data: UInt8ClampedArray = get_image_data(image_path)
* const new_image_data: UInt8ClampedArray = encode_data(image_data, binary_message)
* const encoded_image_path: String = getEncodedImageURI(new_image_data)
```

Let's follow through a worked example (using LSB)

```typescript
* // Input: image_path = "example_white_sqaure.png", message = "A", metadata: {algorithm: "LSB"}

* const compress_message = "A"
* const compressed_binary_message = "0000000101100001"
* const binary_metadata = "00000000"
* const binary_message = "000000000000000101100001"
* const image_data = [255, 255, 255, 255, 255, 255, 255, ..., 255] // 24 `255`'s'
* const new_image_data: UInt8ClampedArray = [254, 254, ... x 13, 255, 254, 255, 255, 254, 254, 254, 254, 255]
* const encoded_image_path = "file:///Stegappasaurus/15011111.png"
```

#### Decoding

All decoding algorithms work as follows

```typescript
* // Input: image_path: String
* // Output: decoded_message: String

* const image_data_buffer = get_image_data() // first 10 pixels
* const metadata: Object = decode_metadata()
* const mesageLength: Number  = decoded_message_length() // will get more pixels from the image if required
* const image_data = get_image_data() // will get the number of pixels we need to decode (start at the correct pixel, not at pixel 0 due to metadata and message length)
* const decoded_decimal_data: Array = decode_data(messageLength, metadata)
* const decoded_message: String = decompress_message(decoded_decimal_data)
```

Let's follow through a worked example (using LSB)

```typescript
* // Input: image_path = "file:///Stegappasaurus/15011111.png"

* const image_data_buffer = [254, 254, ... x 13, 255, 254, 255, 255, 254, 254, 254, 254, 255]
* const metadata = {algorithm: "LSB"}
* const mesageLength = 1
* const image_data = [255, 254, 255, 255, 254, 254, 254, 254, 255]
* const decoded_decimal_data = [65]
* const decoded_message = "A"
```

### Least Significant Bit (LSB)

LSB involves changing the least significant bit, this is because it makes a smaller difference
to the image so it's much harder to spot visually. For example say we want to encode the `011`
into an image and the next pixel has RGB values of RGB(255, 255, 255).

```text
# Most Significant Bit

- Encode `0` into `255`
- Convert `255` into binary = `11111111`
- Result `011111111` = 127
- Pixel = RGB(127, 255, 255)

# Least Significant Bit

- Encode `0` into `255`
- Convert `255` into binary = `11111111`
- Result `111111110` = 254
- Pixel =  RGB(254, 255, 255)

# We don't have to change the other values as 255 is already encoded with a 1 at the beginning.
```

Now our encoded pixel has gone from white to a very light blue, which would be very noticeable.
Now let's try the same example with LSB. However, with the LSB example, it's just a slightly dark white.

Now, how about decoding, well it's relatively simple if the decimal number is even the LSB
is `0` and if it's odd the LSB will always be `1`. Therefore from RGB(254, 255, 255) we decode
`101`.

```text
# Least Significant Bit

- Convert `254` into binary = `11111110`
- Decoded Bit `0`
- Convert `255` into binary = `11111111`
- Decoded Bit `1`
- Convert `255` into binary = `11111111`
- Decoded Bit `1`

- Decoded `101`
```

#### EncodeLSB

The `EncodeLSB` class is very simple, it receives a message as a binary string and the image
data to encode. It will then encode the message into the image data bit by bit. Each element
in the image data will have encoded 1 bit from the message. Sometimes the LSB will not change
for example if we want to encode `1` into `255` since it already ends in a `1` because `255`
is `1111111` in binary we don't have to change the value.

It also calls an action function, every time is encodes a bit, which can be passed to the object
constructor such as incrementing a progress bar.

#### DecodeLSB

The `DecodeLSB` class is also very simple. It decodes one byte at a time because to decode the
image from binary to a normal UTF8 string we need the. To retrieve the image we again go byte
by byte but this time we use a remainder (modulus, %). Essentially if it's odd (must end in a `1`)
the LSB will be `1` and if it's even the LSB must be `0`. We decode 8 bits (a byte) then
push this on to an array. This array will then be decoded and decompressed by the Steganography
class.
