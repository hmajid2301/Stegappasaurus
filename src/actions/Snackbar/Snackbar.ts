import Snackbar from 'react-native-snackbar';

import {primary, pureWhite} from '~/constants/colors';

interface Show {
  text: string;
  buttonText?: string;
  onButtonPress?: () => void;
  duration?: number;
}

export default class CustomSnackbar {
  public static show({
    onButtonPress = () => null,
    buttonText = 'Okay',
    duration = 5000,
    text,
  }: Show) {
    Snackbar.show({
      action: {
        color: primary,
        onPress: onButtonPress,
        title: buttonText,
      },
      color: pureWhite,
      duration,
      title: text,
    });
  }
}
