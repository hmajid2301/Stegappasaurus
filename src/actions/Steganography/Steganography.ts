import LZUTF8 from 'lzutf8';
import {stringToArray} from 'utf8-to-bytes';
import varint from 'varint';

import Canvas, {Image as CanvasImage, ImageData} from 'react-native-canvas';
import {ImageNotEncodedError, MessageTooLongError} from './exceptions';
import {DecodeLSB, EncodeLSB} from './LSB';

type AlgorithmNames = 'LSBv1';

interface IMetaData {
  [name: string]: number | string;
}

interface IAlgorithms {
  [name: number]: AlgorithmNames;
}

export default class Steganography {
  private readonly canvas: Canvas;
  private readonly imageURI: string;
  private readonly width: number;
  private readonly height: number;
  private progress: {
    increment: number;
    value: number;
  };

  private numToAlgorithmTypes: IAlgorithms = {
    0: 'LSBv1',
  };

  constructor(canvas: Canvas, imageURI: string, width: number, height: number) {
    this.canvas = canvas;
    this.imageURI = imageURI;
    this.width = width;
    this.height = height;
    this.canvas.width = width;
    this.canvas.height = height;
    this.progress = {
      increment: 0,
      value: 0,
    };
  }

  public async encode(
    message: string,
    algorithm: AlgorithmNames = 'LSBv1',
    metadata?: IMetaData,
  ) {
    try {
      const output = LZUTF8.compress(message);
      const binaryMessage = this.convertMessageToBits(output);
      const imageData = await this.getImageData(
        this.imageURI,
        binaryMessage.length,
      );
      const newImageData = this.encodeData(
        imageData,
        binaryMessage,
        algorithm,
        metadata,
      );
      const url = await this.getEncodedImageURI(newImageData);
      return url;
    } catch (error) {
      if (error instanceof RangeError) {
        throw new MessageTooLongError('Message too long to encode', message);
      } else {
        throw new Error(error);
      }
    }
  }

  public async decode() {
    try {
      const imageData = await this.getImageData(
        this.imageURI,
        this.width * this.height,
      );
      const decodedDecimalData = this.decodeData(imageData);
      const message = LZUTF8.decompress(decodedDecimalData);
      return message;
    } catch (error) {
      if (error instanceof RangeError) {
        throw new ImageNotEncodedError(
          'Image is not encoded with any data',
          this.imageURI,
        );
      } else {
        throw new Error(error);
      }
    }
  }

  public updateProgress() {
    this.progress.value += this.progress.increment;
  }

  public getProgress() {
    return this.progress.value;
  }

  private convertMessageToBits(message: Uint8Array) {
    const messageLength = varint.encode(message.length);
    const arrayToEncode = [...messageLength, ...message];

    let bitsToEncode: string = '';
    for (const element of arrayToEncode) {
      bitsToEncode += this.convertDecimalToBytes(element);
    }

    return bitsToEncode;
  }

  private async getImageData(imageURI: string, length: number) {
    const context = this.canvas.getContext('2d');
    const uri = imageURI;
    const image = await this.loadImage(uri);
    context.drawImage(image, 0, 0, this.width, this.height);

    const bytesRequired = this.getBytesRequired(length);
    const width = bytesRequired % this.width;
    const height = Math.floor(bytesRequired / this.height) + 1;

    const rgbaImageObject = await context.getImageData(0, 0, width, height);
    const rgbData = Object.values(rgbaImageObject.data).filter((_, index) => {
      return (index + 1) % 4 !== 0;
    });
    return rgbData;
  }

  private loadImage(url: string): Promise<CanvasImage> {
    return new Promise(resolve => {
      const image = new CanvasImage(this.canvas);
      image.addEventListener('load', () => {
        resolve(image);
      });
      image.src = url;
    });
  }

  private getBytesRequired(messageLength: number, _: AlgorithmNames = 'LSBv1') {
    const bytesRequired = Math.ceil((messageLength + 8) / 3);
    return bytesRequired;
  }

  private encodeData(
    imageData: number[],
    data: string,
    algorithm: AlgorithmNames,
    metadata?: IMetaData,
  ) {
    this.progress.increment = 100 / data.length;
    metadata = this.setDefaultMetaData(
      metadata && algorithm !== 'LSBv1' ? metadata : {},
      algorithm,
    );
    const {newImageData, startEncodingAt} = this.encodeMetadata(
      imageData,
      metadata,
    );

    let encodedData;
    switch (algorithm) {
      default: {
        encodedData = new EncodeLSB(this.updateProgress.bind(this)).encode(
          newImageData,
          data,
          startEncodingAt,
        );
      }
    }

    return encodedData;
  }

  private setDefaultMetaData(metadata: IMetaData, algorithm: AlgorithmNames) {
    const algorithmNum = this.algorithmsTypesToNum(algorithm);
    metadata.algorithm = algorithmNum;

    return metadata;
  }

  private algorithmsTypesToNum(algorithm = 'LSB') {
    const values = Object.values(this.numToAlgorithmTypes);
    return values.indexOf(algorithm);
  }

  private encodeMetadata(imageData: number[], metadata: IMetaData) {
    const metadataToEncode = [];
    const dataToEncode = ['algorithm'];

    for (const data of dataToEncode) {
      const tempData = metadata[data];
      if (!tempData && tempData !== 0) {
        continue;
      }

      let decimalDataToEncode;
      if (typeof tempData === 'string') {
        decimalDataToEncode = stringToArray(tempData) as number[];
      } else {
        decimalDataToEncode = varint.encode(tempData);
      }
      metadataToEncode.push(...decimalDataToEncode);
    }

    let binaryMetadata = '';
    for (const data of metadataToEncode) {
      binaryMetadata += this.convertDecimalToBytes(data);
    }
    const imageDataWithMetadata = new EncodeLSB().encode(
      imageData,
      binaryMetadata,
    );

    return {
      newImageData: imageDataWithMetadata,
      startEncodingAt: binaryMetadata.length,
    };
  }

  private async getEncodedImageURI(data: number[]) {
    const context = this.canvas.getContext('2d');
    for (let index = 3; index < data.length + 3; index += 4) {
      data.splice(index, 0, 255);
    }
    const width = data.length / 4;
    const height = Math.floor(width / this.height) + 1;

    const imgData = new ImageData(this.canvas, data, width, height);
    context.putImageData(imgData, 0, 0);
    return this.canvas.toDataURL();
  }

  private decodeData(imageData: number[]) {
    const {algorithm, startDecodingAt} = this.decodeMetadata(imageData);
    let decodedMessage;

    switch (algorithm) {
      default: {
        decodedMessage = new DecodeLSB(this.updateProgress.bind(this)).decode(
          imageData,
          startDecodingAt,
        );
      }
    }

    const decodedDecimal: number[] = [];
    for (const byte of decodedMessage) {
      const decimal = this.convertBytesToDecimal(byte);
      decodedDecimal.push(decimal);
    }

    return new Uint8Array(decodedDecimal);
  }

  private decodeMetadata(imageData: number[]) {
    const decodeLSB = new DecodeLSB();
    const algorithmTypeBinary = decodeLSB.decodeNextByte(imageData);
    const algorithmNum = this.convertBytesToDecimal(algorithmTypeBinary);
    const algorithm = this.numToAlgorithmTypes[algorithmNum] || 'LSB';
    const metadata: IMetaData = {};

    const startDecodingAt = decodeLSB.getCurrentIndex();
    return {
      algorithm,
      metadata,
      startDecodingAt,
    };
  }

  private convertDecimalToBytes(data: number) {
    const binaryString = data.toString(2);
    const nearestByte = Math.ceil(binaryString.length / 8) * 8;
    const byte = binaryString.padStart(nearestByte, '0');
    return byte;
  }

  private convertBytesToDecimal(bytes: string) {
    const decimal = parseInt(bytes, 2);
    return decimal;
  }
}
