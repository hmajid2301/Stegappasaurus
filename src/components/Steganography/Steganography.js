import { Component } from 'react';
import PropTypes from 'prop-types';

export default class Steganography extends Component {
  static propTypes = {
    webview: PropTypes.shape({
      webview: PropTypes.func.isRequired,
    }).isRequired,
  };

  convertMessageToUnicode(message) {
    const unicodeMessage = [];
    message.forEach((character) => {
      const unicode = character.codePointAt(0);
      unicodeMessage.push(unicode);
    });

    return unicodeMessage;
  }

  convertUnicodeToMessage(unicodeList) {
    let stringMessage = '';
    unicodeList.forEach((unicode) => {
      stringMessage += String.fromCodePoint(unicode);
    });

    return stringMessage;
  }

  getPixelData() {
    this.webview.postMessage(JSON.stringify({ name: 'getPixelData', args: '' }));
  }

  setPixelData(pixelData) {
    this.webview.postMessage(JSON.stringify({ name: 'setPixelData', args: pixelData }));
  }
}
