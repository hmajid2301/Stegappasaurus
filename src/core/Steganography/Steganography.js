import PropTypes from 'prop-types';

export default class Steganography {
  static propTypes = {
    webview: PropTypes.shape({
      webview: PropTypes.func.isRequired,
    }).isRequired,
  };

  static getSeparator() {
    return '1101100010100111';
  }

  convertMessageToBinary(message) {
    const unicodeMessage = [];
    const messageLength = message.length;
    const messageSizeBinary = messageLength.toString(2);

    unicodeMessage.push(messageSizeBinary);
    unicodeMessage.push(this.getSeparator());

    message.forEach((character) => {
      const unicode = character.codePointAt(0);
      let binaryValue = unicode.toString(2);
      if (binaryValue.length < 8) {
        const padLength = 8 - (binaryValue.length - 1);
        binaryValue = binaryValue.padStart(padLength, '0');
      }

      unicodeMessage.push(binaryValue);
    });

    return unicodeMessage;
  }

  convertUnicodeToMessage(unicodeList) {
    let stringMessage = '';
    unicodeList.forEach((unicode) => {
      stringMessage += String.fromCodePoint(unicode);
    });

    return stringMessage;
  }

  getPixelData() {
    this.webview.postMessage(JSON.stringify({ name: 'getPixelData', args: '' }));
  }

  setPixelData(pixelData) {
    this.webview.postMessage(JSON.stringify({ name: 'setPixelData', args: pixelData }));
  }
}
