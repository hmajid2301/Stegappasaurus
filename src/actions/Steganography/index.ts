import Steganography, {Actions as StegActions} from './Steganography';

export type Actions = StegActions;
export default Steganography;
