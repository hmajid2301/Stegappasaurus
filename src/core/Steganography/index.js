import DCT from './DCT';
import Common from './Common';
import Steganography from './Steganography';


export {
  DCT,
  Common,
  Steganography,
};
