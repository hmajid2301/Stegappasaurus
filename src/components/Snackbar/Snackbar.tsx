import Snackbar from "react-native-snackbar";

import { colors } from "~/constants";

interface IShow {
  text: string;
  buttonText?: string;
  onButtonPress?: () => void;
  duration?: number;
}

export default class CustomSnackbar {
  public static show = ({
    onButtonPress = () => null,
    buttonText = "Okay",
    duration = 5000,
    text
  }: IShow) => {
    Snackbar.show({
      action: {
        color: colors.primary,
        onPress: onButtonPress,
        title: buttonText
      },
      duration,
      title: text
    });
  };
}
