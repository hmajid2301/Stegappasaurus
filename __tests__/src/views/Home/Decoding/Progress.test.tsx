import {render, wait} from '@testing-library/react-native';
import React from 'react';

import Steganography from '~/actions/Steganography';
import Snackbar from '~/actions/Snackbar';
import Progress from '~/views/Home/Decoding/Progress';

jest.mock('~/actions/Steganography');

const navigation: any = {
  addListener: jest.fn(),
  navigate: jest.fn(),
  getParam: jest.fn().mockReturnValue({
    uri: 'file:///storage/emulated/0/Stegappasaurus/1575927505003.png',
  }),
  state: {
    routeName: 'Progress',
  },
};

describe('Decoding Progress: Functionality', () => {
  test('Success', async () => {
    Steganography.prototype.decode = jest.fn().mockResolvedValue('Hello!');

    const spy = jest.spyOn(navigation, 'navigate');

    render(<Progress navigation={navigation} />);

    await wait(() => {
      expect(spy).toHaveBeenCalledWith('Message', {
        message: 'Hello!',
        uri: {
          uri: 'file:///storage/emulated/0/Stegappasaurus/1575927505003.png',
        },
      });
    });
  });

  test('Failed', async () => {
    Steganography.prototype.decode = jest.fn().mockImplementation(() => {
      throw new Error();
    });

    const snackSpy = jest.spyOn(Snackbar, 'show');
    const navSpy = jest.spyOn(navigation, 'navigate');

    render(<Progress navigation={navigation} />);

    await wait(() => {
      expect(snackSpy).toHaveBeenCalledWith({
        text: 'Failed to decode image, please try again.',
      });

      expect(navSpy).toHaveBeenCalledWith('Main');
    });
  });
});
