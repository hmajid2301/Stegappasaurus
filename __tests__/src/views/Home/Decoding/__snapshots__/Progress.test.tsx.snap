// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decoding Progress: Match Snapshots 1 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ImageProgress
    animating={true}
    background="#FFF"
    primaryColor="#E88C0C"
  />
</View>
`;

exports[`Decoding Progress: Match Snapshots 2 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ImageProgress
    animating={true}
    background="#17212D"
    primaryColor="#E88C0C"
  />
</View>
`;
